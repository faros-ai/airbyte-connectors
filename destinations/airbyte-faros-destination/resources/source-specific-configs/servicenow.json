{
  "title": "ServiceNow",
  "description": "Configuration options that apply to records generated by the ServiceNow Source.",
  "type": "object",
  "oneOf": [
    {
      "type": "object",
      "title": "Configuration",
      "properties": {
        "source_type": {
          "type": "string",
          "const": "ServiceNow",
          "order": 0
        },
        "application_mapping": {
          "type": "string",
          "title": "Application Mapping",
          "description": "JSON map of ServiceNow service(s) name, to compute platform specific app name and platform name. Used to reference compute_Application object, from an ims_Incident object.",
          "multiline": true,
          "default": "{}",
          "examples": [
            "{ \"Aion\": { \"name\": \"aion\", \"platform\": \"ECS\" } }"
          ]
        },
        "application_field": {
          "type": "string",
          "title": "Application Field",
          "description": "Field name in a ServiceNow incident, to compute a Faros Application name. The value of this field will be used to look up the app/platform in the specified application mapping",
          "default": "business_service",
          "pattern": "^(business_service|cmdb_ci)$",
          "examples": [
            "business_service",
            "cmdb_ci"
          ]
        },
        "default_severity": {
          "type": "string",
          "title": "Default Severity",
          "description": "A default severity category if not present",
          "pattern": "^(Sev[1-5])?(Custom)?$",
          "examples": [
            "Sev1",
            "Sev2",
            "Sev3",
            "Sev4",
            "Sev5",
            "Custom"
          ]
        },
        "default_priority": {
          "type": "string",
          "title": "Default Priority",
          "description": "A default priority category if not present",
          "pattern": "^(Critical|High|Medium|Low|Custom)$",
          "examples": [
            "Critical",
            "High",
            "Medium",
            "Low",
            "Custom"
          ]
        },
        "store_current_incidents_associations": {
          "type": "boolean",
          "title": "Only store current incidents associations",
          "description": "Only store current incidents associations in the graph.",
          "default": false
        }
      }
    }
  ]
}
