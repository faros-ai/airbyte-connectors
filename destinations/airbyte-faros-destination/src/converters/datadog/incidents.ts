import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {Common} from '../common/common';
import {DestinationModel, DestinationRecord, StreamContext} from '../converter';
import {DatadogConverter, IncidentSeverityCategory} from './common';

enum IncidentStatusCategory {
  Created = 'Created',
  Identified = 'Identified',
  Investigating = 'Investigating',
  Resolved = 'Resolved',
  Custom = 'Custom',
}

export class Incidents extends DatadogConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'ims_Incident',
    'ims_IncidentAssignment',
  ];

  private seenApplications = new Set<string>();

  async convert(
    record: AirbyteRecord,
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const config = this.config(ctx);
    const source = this.streamName.source;
    const incident = record.record.data;
    const incidentKey = {
      uid: incident.id,
      source,
    };

    const defaultSeverityCategory = config.default_severity;
    const defaultSeverity = defaultSeverityCategory
      ? {
          category: defaultSeverityCategory,
          detail: 'default',
        }
      : null;

    const res = [];
    res.push({
      model: 'ims_Incident',
      record: {
        ...incidentKey,
        title: incident.attributes?.title,
        description: incident.attributes?.fields?.summary?.value,
        url: null,
        severity:
          this.getSeverity(incident.attributes?.fields?.severity?.value) ??
          defaultSeverity,
        priority: null,
        status:
          this.getStatus(incident.attributes?.fields?.state?.value) ?? null,
        createdAt: Utils.toDate(incident.attributes?.created) ?? null,
        updatedAt: Utils.toDate(incident.attributes?.modified) ?? null,
        resolvedAt: Utils.toDate(incident.attributes?.resolved) ?? null,
      },
    });

    const assigneeUid = incident.relationships?.commanderUser?.data?.id;
    if (assigneeUid) {
      res.push({
        model: 'ims_IncidentAssignment',
        record: {
          incident: incidentKey,
          assignee: {
            uid: assigneeUid,
            source,
          },
        },
      });
    }

    const services: string[] = incident.attributes?.fields?.services?.value;
    if (services) {
      const applicationMapping = this.applicationMapping(ctx);
      for (const service of services) {
        if (!service) continue;
        const mappedApp = applicationMapping[service];
        const application = Common.computeApplication(
          mappedApp?.name ?? service,
          mappedApp?.platform
        );
        const appKey = application.uid;
        if (!this.seenApplications.has(appKey)) {
          res.push({model: 'compute_Application', record: application});
          this.seenApplications.add(appKey);
        }
        res.push({
          model: 'ims_IncidentApplicationImpact',
          record: {incident: incidentKey, application},
        });
      }
    }

    return res;
  }

  private getSeverity(
    severity?: string
  ): {category: IncidentSeverityCategory; detail: string} | undefined {
    if (severity) {
      switch (severity) {
        case 'SEV-1':
          return {category: IncidentSeverityCategory.Sev1, detail: severity};
        case 'SEV-2':
          return {category: IncidentSeverityCategory.Sev2, detail: severity};
        case 'SEV-3':
          return {category: IncidentSeverityCategory.Sev3, detail: severity};
        case 'SEV-4':
          return {category: IncidentSeverityCategory.Sev4, detail: severity};
        case 'SEV-5':
          return {category: IncidentSeverityCategory.Sev5, detail: severity};
        default:
          return {category: IncidentSeverityCategory.Custom, detail: severity};
      }
    }
  }

  private getStatus(
    state?: string
  ): {category: IncidentStatusCategory; detail: string} | undefined {
    if (state) {
      switch (state) {
        case 'declared':
          return {category: IncidentStatusCategory.Created, detail: state};
        case 'active':
          return {
            category: IncidentStatusCategory.Investigating,
            detail: state,
          };
        case 'stable':
          return {category: IncidentStatusCategory.Identified, detail: state};
        case 'resolved':
          return {category: IncidentStatusCategory.Resolved, detail: state};
        default:
          return {category: IncidentStatusCategory.Custom, detail: state};
      }
    }
  }
}
