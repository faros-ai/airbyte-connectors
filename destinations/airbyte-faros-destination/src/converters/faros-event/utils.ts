import {ErrorObject} from 'ajv';
import {groupBy} from 'lodash';

import {Artifact, FarosPath, FarosTag, RunStep} from './types';

interface OrganizationKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeploymentKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeployIds {
  readonly id: string;
  readonly source: string;
  readonly environment: string;
  readonly application: string;
}

export interface ArtifactKey {
  readonly uid: string;
  readonly repository: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface ArtifactIds {
  readonly id: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface CommitKey {
  readonly sha: string;
  readonly repository: {
    readonly name: string;
    readonly organization: OrganizationKey;
  };
}

export interface CommitIds {
  readonly sha: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildKey {
  readonly uid: string;
  readonly pipeline: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface BuildIds {
  readonly id: string;
  readonly pipeline: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildStepKey {
  readonly uid: string;
  readonly build: BuildKey;
}

export interface AgentKey {
  readonly uid: string;
  readonly source: string;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveCommitIds(commit: any): CommitIds {
  if (commit.uri) {
    const parsedUri = parseUri(commit.uri);
    return {
      sha: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    sha: commit.sha,
    repository: commit.repository.toLowerCase(),
    organization: commit.organization.toLowerCase(),
    source: commit.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeCommitKey(commit: any): CommitKey {
  const commitIds = resolveCommitIds(commit);
  return {
    sha: commitIds.sha,
    repository: {
      name: commitIds.repository,
      organization: {
        uid: commitIds.organization,
        source: commitIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveBuildIds(run: any): BuildIds {
  if (run.uri) {
    const parsedUri = parseUri(run.uri);
    return {
      id: parsedUri[3],
      pipeline: parsedUri[2],
      organization: parsedUri[1],
      source: parsedUri[0],
    };
  }
  return {
    id: run.id,
    pipeline: run.pipeline,
    organization: run.organization,
    source: run.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeBuildKey(run: any): BuildKey {
  const buildIds = resolveBuildIds(run);
  return {
    uid: buildIds.id,
    pipeline: {
      uid: buildIds.pipeline,
      organization: {
        uid: buildIds.organization,
        source: buildIds.source,
      },
    },
  };
}

export function makeBuildStepKey(step: RunStep, build: BuildKey): BuildStepKey {
  return {
    uid: step.id,
    build,
  };
}

export function resolveDeploymentIds(deploy: any): DeployIds {
  let deployIds;
  if (deploy.uri) {
    const parsedUri = parseUri(deploy.uri);
    deployIds = {
      id: parsedUri[3],
      environment: parsedUri[2],
      application: parsedUri[1],
      source: parsedUri[0],
    };
  } else {
    deployIds = {
      id: deploy.id,
      environment: deploy.environment,
      application: deploy.application,
      source: deploy.source,
    };
  }
  return deployIds;
}

function enumToMap(enumValues: string[]): Map<string, string> {
  return new Map(enumValues.map((value) => [value.toLowerCase(), value]));
}

function resolveEnum(
  enumValue: Map<string, string>,
  field: string,
  fieldDetail?: string | null
): {category: string; detail?: string | null} {
  const lowerField = field.toLowerCase();
  const value = enumValue.get(lowerField);
  if (value) {
    return {category: value, detail: fieldDetail};
  }
  return {
    category: 'Custom',
    detail: field.concat(fieldDetail ? ` - ${fieldDetail}` : ''),
  };
}

const deploymentEnvironments = enumToMap([
  'Prod',
  'Staging',
  'QA',
  'Dev',
  'Sandbox',
  'Canary',
  'Custom',
]);

export function resolveDeploymentEnvironment(
  environment: string,
  environmentDetails?: string | null
): {category: string; detail?: string | null} {
  return resolveEnum(deploymentEnvironments, environment, environmentDetails);
}

const deploymentStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'RolledBack',
  'Success',
]);

export function resolveDeploymentStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(deploymentStatuses, status, statusDetails);
}

const runStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'Success',
  'Unknown',
]);

export function resolveRunStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(runStatuses, status, statusDetails);
}

const runStepTypes = enumToMap(['Custom', 'Manual', 'Script']);

export function resolveRunStepType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(runStepTypes, type, typeDetails);
}

const computeAgentOS = enumToMap(['Linux', 'MacOS', 'Windows', 'Custom']);

export function resolveComputeAgentOS(
  os?: string,
  osDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!os) {
    return undefined;
  }
  return resolveEnum(computeAgentOS, os, osDetails);
}

const metricDefinitionTypes = enumToMap([
  'Numeric',
  'String',
  'Date',
  'Timestamp',
  'Duration',
  'Percentage',
  'Custom',
]);

export function resolveMetricDefinitionType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(metricDefinitionTypes, type, typeDetails);
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveArtifactIds(artifact: any): ArtifactIds {
  if (artifact.uri) {
    const parsedUri = parseUri(artifact.uri);
    return {
      id: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    id: artifact.id,
    repository: artifact.repository.toLowerCase(),
    organization: artifact.organization.toLowerCase(),
    source: artifact.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeArtifactKey(artifact: any): ArtifactKey {
  const artifactIds = resolveArtifactIds(artifact);
  return {
    uid: artifactIds.id,
    repository: {
      uid: artifactIds.repository.toLowerCase(),
      organization: {
        uid: artifactIds.organization.toLowerCase(),
        source: artifactIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeDummyArtifact(commit: any): Artifact {
  const commitKey = makeCommitKey(commit);
  return {
    id: commitKey.sha,
    repository: commitKey.repository.name,
    organization: commitKey.repository.organization.uid,
    source: commitKey.repository.organization.source,
  };
}

// Convert ISO-8601 string, number string, number, or 'Now' into milliseconds
export function getTime(date?: string | number): number | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return Date.now();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).getTime();
      }
    }
    return new Date(date).getTime();
  }
  return undefined;
}

// Convert ISO-8601 string, number string, number, or 'Now' into ISO-8601 string
export function toDate(date?: string | number): string | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return new Date(Date.now()).toISOString();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).toISOString();
      }
    }
    return new Date(date).toISOString();
  }
  return undefined;
}

export function parseTags(tags?: FarosTag[] | string): FarosTag[] | undefined {
  const parser = (input: string): FarosTag[] => {
    const resTags: FarosTag[] = [];
    const keyValues = input.split(',');
    for (const keyValue of keyValues) {
      const splitTag = keyValue.split(':');
      resTags.push({key: splitTag[0], value: splitTag[1]});
    }
    return resTags;
  };
  return parseArray(parser, tags);
}

export function parsePaths(
  paths?: FarosPath[] | string
): FarosPath[] | undefined {
  const parser = (input: string): FarosPath[] => {
    return input.split(',').map<FarosPath>((path) => {
      return {path};
    });
  };
  return parseArray(parser, paths);
}

export function parsePullRequestNumber(
  number: string | number
): number | undefined {
  const prNum = typeof number === 'string' ? Number(number) : number;
  return isNaN(prNum) ? undefined : prNum;
}

export function parseUri(uri: string): string[] {
  const components: string[] = [];
  const split1 = uri.split('://');
  components.push(split1[0]);
  split1[1].split('/').forEach((value) => components.push(value));
  return components;
}

export function parseStringArray(
  array?: string[] | string,
  separator = ','
): string[] | undefined {
  return parseArray((s) => s.split(separator), array);
}

function parseArray<T>(
  parser: (input: string) => T[],
  input?: T[] | string
): T[] | undefined {
  if (typeof input === 'string') {
    return parser(input);
  }
  return input;
}

export interface EventValidationError {
  readonly message: string;
  readonly details?: ReadonlyArray<string>;
  readonly allowedValues?: ReadonlyArray<string>;
  readonly additionalProperties?: ReadonlyArray<string>;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function removeUndefinedProperties(object: any): any {
  const result = {...object};
  Object.keys(result).forEach(
    (key) => result[key] === undefined && delete result[key]
  );
  return result;
}

export function formatErrors(
  errors?: ErrorObject[] | null,
  rootObject?: string
): ReadonlyArray<EventValidationError> {
  if (!errors) {
    return [];
  }

  const formattedErrors: EventValidationError[] = [];
  const groupedErrors = groupBy<ErrorObject>(errors, (err) => {
    return formatErrorDetailsMessage(err, rootObject);
  });

  for (const [message, errs] of Object.entries(groupedErrors)) {
    const details = errs.flatMap(
      (err) =>
        err.params?.errors?.map((v: any) =>
          formatErrorDetailsMessage(v, rootObject)
        ) ?? []
    );
    const additionalProperties = errs.flatMap(
      (err) => err.params?.additionalProperty ?? []
    );
    formattedErrors.push({
      message,
      details: details.length > 0 ? details : undefined,
      allowedValues: getAllowedValues(errs),
      additionalProperties:
        additionalProperties.length > 0 ? additionalProperties : undefined,
    });
  }
  return formattedErrors;
}

function formatErrorDetailsMessage(
  error: ErrorObject,
  rootObject = 'data'
): string {
  const prefix = error?.instancePath
    ? `${rootObject}${error.instancePath.replace(/\//g, '.')} `
    : `${rootObject} `;
  return `${prefix}${error.message}`;
}

function getAllowedValues(errors: ErrorObject[]): string[] | undefined {
  let result;
  if (errors?.length > 0) {
    result = errors[0].params?.allowedValues;
    if (!result && errors[0].params?.errors?.length > 0) {
      result = errors[0].params?.errors[0].params?.allowedValues;
    }
  }
  return result;
}

// Max length for free-form description text fields such as issue body
const MAX_TEXT_LENGTH = 1000;
export function cleanAndTruncate(
  str?: string,
  maxLength?: number
): string | null | undefined {
  if (!str) {
    return str;
  }
  const length = maxLength ?? MAX_TEXT_LENGTH;
  let result: string;
  if (str.length <= length) {
    result = str;
  } else {
    // If the last character is part of a unicode surrogate pair, include the next character
    const lastChar = str.codePointAt(length - 1) ?? 0;
    result =
      lastChar > 65535
        ? str.substring(0, length + 1)
        : str.substring(0, length);
  }
  // eslint-disable-next-line no-control-regex
  return result.replace(/\u0000/g, '');
}
