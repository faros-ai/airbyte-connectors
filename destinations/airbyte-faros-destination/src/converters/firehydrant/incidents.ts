import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {Common} from '../common/common';
import {DestinationModel, DestinationRecord, StreamContext} from '../converter';
import {FireHydrantConverter} from './common';
import {IncidentTicket} from './models';
import {
  FirehydrantIncidentMilestone,
  FirehydrantIncidentPriority,
  FirehydrantIncidentSeverity,
  Incident,
  IncidentEventType,
  IncidentEventTypeCategory,
  IncidentPriority,
  IncidentPriorityCategory,
  IncidentSeverity,
  IncidentSeverityCategory,
  IncidentStatusCategory,
  IncidentTicketState,
  TaskStatus,
  TaskStatusCategory,
} from './models';

export class Incidents extends FireHydrantConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'compute_Application',
    'ims_Incident',
    'ims_IncidentApplicationImpact',
    'ims_IncidentAssignment',
    'ims_IncidentEvent',
    'ims_IncidentTag',
    'ims_IncidentTasks',
    'ims_Label',
    'tms_Task',
  ];

  private seenTags = new Set<string>();
  private seenApplications = new Set<string>();

  async convert(
    record: AirbyteRecord,
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const incident = record.record.data as Incident;
    const res: DestinationRecord[] = [];

    const maxDescriptionLength = this.maxDescriptionLength(ctx);

    const incidentRef = {uid: incident.id, source};
    const createdAt = Utils.toDate(incident.created_at);
    const updatedAt = incident.events
      ? Utils.toDate(incident.events[incident.events.length - 1].occurred_at)
      : createdAt;

    let acknowledgedAt: Date = undefined;
    let resolvedAt: Date = undefined;

    for (const event of incident.events) {
      const eventType: IncidentEventType = {
        category: IncidentEventTypeCategory.Created,
        detail: event.type,
      };

      const occurredAt = Utils.toDate(event.occurred_at);
      if (
        !resolvedAt &&
        event.data.current_milestone === FirehydrantIncidentMilestone.resolved
      ) {
        resolvedAt = occurredAt;
        eventType.category = IncidentEventTypeCategory.Resolved;
      }
      if (
        !acknowledgedAt &&
        event.data.current_milestone ===
          FirehydrantIncidentMilestone.acknowledged
      ) {
        acknowledgedAt = occurredAt;
        eventType.category = IncidentEventTypeCategory.Acknowledged;
      }
      res.push({
        model: 'ims_IncidentEvent',
        record: {
          uid: event.id,
          type: eventType,
          incident: incidentRef,
          detail: JSON.stringify(event.data),
          createdAt: occurredAt,
        },
      });

      if (event.data.operation === 'created' && event.data.ticket) {
        const ticket = event.data.ticket as IncidentTicket;
        // TODO
        res.push(
          this.getTaskDestinationRecord(
            ticket,
            source,
            maxDescriptionLength,
            occurredAt
          )
        );
        const task = {uid: ticket.id, source};
        res.push({
          model: 'ims_IncidentTasks',
          record: {
            task,
            incident: incidentRef,
          },
        });
      }
    }

    res.push({
      model: 'ims_Incident',
      record: {
        ...incidentRef,
        title: incident.name,
        description: incident.description?.substring(0, maxDescriptionLength),
        url: incident.incident_url,
        createdAt,
        updatedAt,
        acknowledgedAt,
        resolvedAt,
        priority: this.getPriority(incident.priority),
        severity: this.getSeverity(incident.severity),
        status: this.getIncidentStatus(incident.current_milestone),
      },
    });

    if (incident.services) {
      const applicationMapping = this.applicationMapping(ctx);
      for (const service of incident.services) {
        if (!service?.name) continue;
        const mappedApp = applicationMapping[service.name];
        const application = Common.computeApplication(
          mappedApp?.name ?? service.name,
          mappedApp?.platform
        );
        const appKey = application.uid;
        if (!this.seenApplications.has(appKey)) {
          res.push({model: 'compute_Application', record: application});
          this.seenApplications.add(appKey);
        }
        res.push({
          model: 'ims_IncidentApplicationImpact',
          record: {incident: incidentRef, application},
        });
      }
    }

    for (const assignment of incident.role_assignments) {
      const assignee = {uid: assignment.user.id, source};
      res.push({
        model: 'ims_IncidentAssignment',
        record: {
          assignee,
          incident: incidentRef,
        },
      });
    }

    for (const tag of incident.tag_list) {
      if (!this.seenTags.has(tag)) {
        this.seenTags.add(tag);
        res.push({
          model: 'ims_Label',
          record: {
            name: tag,
          },
        });
      }
      res.push({
        model: 'ims_IncidentTag',
        record: {
          label: {name: tag},
          incident: incidentRef,
        },
      });
    }
    return res;
  }

  private getTaskDestinationRecord(
    ticket: IncidentTicket,
    source: string,
    maxDescriptionLength: number,
    occurredAt: Date
  ): DestinationRecord {
    return {
      model: 'tms_Task',
      record: {
        uid: ticket.id,
        name: ticket.summary,
        description: ticket.description?.substring(0, maxDescriptionLength),
        source,
        url: undefined,
        type: {
          detail: ticket.type,
          category: 'Task',
        },
        priority: undefined,
        status: this.getTaskStatus(ticket.state),
        points: 0,
        additionalFields: [],
        createdAt: occurredAt,
        updatedAt: occurredAt,
        statusChangedAt: undefined,
        statusChangelog: undefined,
        parent: undefined,
        creator: {uid: ticket.created_by.id, source},
        epic: undefined,
        sprint: undefined,
      },
    };
  }
  private getPriority(priority: string): IncidentPriority {
    const detail: string = priority;
    switch (priority) {
      case FirehydrantIncidentPriority.P1:
        return {category: IncidentPriorityCategory.Critical, detail};
      case FirehydrantIncidentPriority.P2:
        return {category: IncidentPriorityCategory.High, detail};
      case FirehydrantIncidentPriority.P3:
        return {category: IncidentPriorityCategory.Medium, detail};
      case FirehydrantIncidentPriority.P4:
        return {category: IncidentPriorityCategory.Low, detail};
      default:
        return {category: IncidentPriorityCategory.Custom, detail};
    }
  }

  private getSeverity(severity: string): IncidentSeverity {
    const detail: string = severity;
    switch (severity) {
      case FirehydrantIncidentSeverity.SEV1:
        return {category: IncidentSeverityCategory.Sev1, detail};
      case FirehydrantIncidentSeverity.SEV2:
        return {category: IncidentSeverityCategory.Sev2, detail};
      case FirehydrantIncidentSeverity.SEV3:
        return {category: IncidentSeverityCategory.Sev3, detail};
      case FirehydrantIncidentSeverity.SEV4:
        return {category: IncidentSeverityCategory.Sev4, detail};
      case FirehydrantIncidentSeverity.SEV5:
        return {category: IncidentSeverityCategory.Sev5, detail};
      default:
        return {category: IncidentSeverityCategory.Custom, detail};
    }
  }

  //https://support.firehydrant.io/hc/en-us/articles/4403969187604-Incident-Milestones
  private getIncidentStatus(milestone: string): {
    category: string;
    detail: string;
  } {
    const detail = milestone;
    switch (milestone) {
      case FirehydrantIncidentMilestone.started:
        return {category: IncidentStatusCategory.Investigating, detail};
      case FirehydrantIncidentMilestone.detected:
        return {category: IncidentStatusCategory.Identified, detail};
      case FirehydrantIncidentMilestone.acknowledged:
      case FirehydrantIncidentMilestone.firstaction:
        return {category: IncidentStatusCategory.Monitoring, detail};
      case FirehydrantIncidentMilestone.mitigated:
      case FirehydrantIncidentMilestone.resolved:
        return {category: IncidentStatusCategory.Resolved, detail};
      default:
        return {category: IncidentStatusCategory.Custom, detail};
    }
  }

  private getTaskStatus(status: string): TaskStatus {
    const detail: string = status;
    switch (status) {
      case IncidentTicketState.open:
        return {category: TaskStatusCategory.Todo, detail};
      case IncidentTicketState.in_progress:
        return {category: TaskStatusCategory.InProgress, detail};
      case IncidentTicketState.done:
        return {category: TaskStatusCategory.Done, detail};
      default:
        return {category: TaskStatusCategory.Custom, detail};
    }
  }
}
