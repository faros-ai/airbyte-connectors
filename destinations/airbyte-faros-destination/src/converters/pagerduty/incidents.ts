import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {Common} from '../common/common';
import {DestinationModel, DestinationRecord, StreamContext} from '../converter';
import {PagerDutyConverter, PagerdutyObject} from './common';

enum IncidentStatusCategory {
  Created = 'Created',
  Identified = 'Identified',
  Investigating = 'Investigating',
  Resolved = 'Resolved',
  Custom = 'Custom',
}

type IncidentUrgency = 'high' | 'low'; //PagerDuty only has these two priorities
type IncidentState = 'triggered' | 'acknowledged' | 'resolved';

interface Acknowledgement {
  at: string; //date-time
  acknowledger: PagerdutyObject;
}

export class Incidents extends PagerDutyConverter {
  private seenApplications = new Set<string>();

  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'compute_Application',
    'ims_Incident',
    'ims_IncidentApplicationImpact',
    'ims_IncidentAssignment',
  ];

  async convert(
    record: AirbyteRecord,
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const incident = record.record.data;
    const res: DestinationRecord[] = [];
    const incidentRef = {uid: incident.id, source};
    const lastUpdated = Utils.toDate(incident.last_status_change_at);

    let acknowledgedAt, resolvedAt;
    if (incident.status === 'acknowledged') {
      if (!incident.acknowledgements || !incident.acknowledgements.length) {
        ctx.logger.warn(
          `Incident ${incident.id} acknowledged, but acknowledger info missing`
        );
      } else {
        // find first acknowledgement
        acknowledgedAt = Utils.toDate(
          incident.acknowledgements
            .map((ack: Acknowledgement) => ack.at)
            .sort()[0]
        );
      }
    } else if (incident.status === 'resolved') {
      resolvedAt = lastUpdated;
    }

    res.push({
      model: 'ims_Incident',
      record: {
        ...incidentRef,
        title: incident.title,
        description: incident.description,
        url: incident.self,
        createdAt: Utils.toDate(incident.created_at),
        updatedAt: resolvedAt,
        acknowledgedAt: acknowledgedAt,
        resolvedAt: resolvedAt,
        priority: this.incidentPriority(incident.urgency),
        status: this.incidentState(incident.status),
      },
    });

    for (const assignment of incident.assignments) {
      const assignee = {uid: assignment.assignee.id, source};
      res.push({
        model: 'ims_IncidentAssignment',
        record: {
          incident: incidentRef,
          assignee,
        },
      });
    }

    if (incident?.service?.summary) {
      const applicationMapping = this.applicationMapping(ctx);
      const mappedApp = applicationMapping[incident.service.summary];
      const application = Common.computeApplication(
        mappedApp?.name ?? incident.service.summary,
        mappedApp?.platform
      );
      const appKey = application.uid;
      if (!this.seenApplications.has(appKey)) {
        res.push({model: 'compute_Application', record: application});
        this.seenApplications.add(appKey);
      }
      res.push({
        model: 'ims_IncidentApplicationImpact',
        record: {incident: incidentRef, application},
      });
    }

    return res;
  }

  private incidentPriority(
    incidentUrgency: IncidentUrgency
  ): Record<string, string> {
    const detail = incidentUrgency;
    switch (incidentUrgency.toLowerCase()) {
      case 'high':
        return {category: 'High', detail};
      case 'low':
        return {category: 'Low', detail};
      default:
        return {category: 'Custom', detail};
    }
  }

  private incidentState(incidentStatus: IncidentState): {
    category: string;
    detail: string;
  } {
    const detail = incidentStatus;
    switch (incidentStatus) {
      case 'resolved':
        return {category: IncidentStatusCategory.Resolved, detail};
      case 'acknowledged':
        return {category: IncidentStatusCategory.Investigating, detail};
      case 'triggered':
      default:
        return {category: IncidentStatusCategory.Created, detail};
    }
  }
}
