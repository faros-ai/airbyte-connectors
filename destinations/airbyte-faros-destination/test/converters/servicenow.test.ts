import {Incidents} from '../../src/converters/servicenow/incidents';
import {generateBasicTestSuite} from './utils';

generateBasicTestSuite({sourceName: 'servicenow'});

describe('incident/application association', () => {
  const incAppImpact = {
    incident: {uid: 'incident123', source: 'source1'},
    application: {name: 'app1', platform: 'platform1'},
  };

  const incAppImpacts = {
    incident123: new Set<string>(['app1_platform1']),
  };

  test('should return false if incident source does not match', () => {
    const result = Incidents.shouldDeleteRecord(
      incAppImpact,
      'differentSource',
      incAppImpacts
    );

    expect(result).toBe(false);
  });

  test('should return false if incident was not processed', () => {
    const result = Incidents.shouldDeleteRecord(
      {incident: {uid: 'unprocessed', source: 'source1'}},
      'source1',
      incAppImpacts
    );

    expect(result).toBe(false);
  });

  test('keeps new incident/application association', () => {
    const result = Incidents.shouldDeleteRecord(
      incAppImpact,
      'source1',
      incAppImpacts
    );

    expect(result).toBe(false);
  });

  test('deletes old associations of processed incidents', () => {
    const result = Incidents.shouldDeleteRecord(
      {
        incident: {uid: 'incident123', source: 'source1'},
        application: {name: 'app0', platform: 'platform0'},
      },
      'source1',
      incAppImpacts
    );

    expect(result).toBe(true);
  });
});
