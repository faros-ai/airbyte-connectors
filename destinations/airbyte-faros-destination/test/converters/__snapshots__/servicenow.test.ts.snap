// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`servicenow basic test process records from all streams 1`] = `
Array [
  "Processed 4 records",
  "Processed records by stream: {\\\\\\"mytestsource__servicenow__incidents\\\\\\":2,\\\\\\"mytestsource__servicenow__users\\\\\\":2}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 9 records",
  "Would write records by model: {\\\\\\"compute_Application\\\\\\":1,\\\\\\"ims_Incident\\\\\\":2,\\\\\\"ims_IncidentApplicationImpact\\\\\\":2,\\\\\\"ims_IncidentAssignment\\\\\\":2,\\\\\\"ims_User\\\\\\":2}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;
